python/ray/serve/master.py

```diff
@@ -311,22 +311,33 @@ async def _start_pending_replicas(self):
 
         Clears self.replicas_to_start.
         """
+
+        async def create_one_replica(backend_tag, replica_tag):
+            # NOTE(edoakes): the replicas may already be created if we
+            # failed after creating them but before writing a
+            # checkpoint.
+            try:
+                worker_handle = ray.util.get_actor(replica_tag)
+            except ValueError:
+                worker_handle = await self._start_backend_worker(
+                    backend_tag, replica_tag)
+
+            self.replicas[backend_tag].append(replica_tag)
+            self.workers[backend_tag][replica_tag] = worker_handle
+
+            # Register the worker with the router.
+            await self.router.add_new_worker.remote(backend_tag, replica_tag,
+                                                    worker_handle)
+
+        # We collect all creation coroutines in this list and await on them
```

Comment:
This is a bit too much asyncio magic for my liking, makes it harder to reason about the behavior (in this case maybe not a big deal, but could be in the future). Can we instead do something similar to my PR but just `asyncio.wait()` for each ready task as it completes and then add the worker to the router? I think that would be a bit easier to read & is more "ray-like."

Reply:
I guess that might be a little tricky ^ because we need to associate the backend & replica tags with the futures. Not sure how easy that is. Thoughts?